### PR TITLE
plan: support DataSource converts to new plan.

### PR DESCRIPTION
**Pull request by hanfei1991**
@shenli @coocood @zimulala PTAL

### CODE REVIEW

File: plan/new_physical_plan_builder.go [hunk 1]
```diff
@@ -145,30 +152,245 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
 	return task, p.storeTaskProfile(prop, task)
 }
 
-// TODO: The behavior of Limit may be same with many other plans. Most plans can pass the props or refuse and enforce it.
-func (p *Limit) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error) {
+func planCanPushDown(p LogicalPlan) bool {
```
Comment (coocood) [inline plan/new_physical_plan_builder.go]: add comments for this function

File: plan/logical_plan_test.go [hunk 2]
```diff
@@ -68,17 +68,17 @@ func mockResolve(node ast.Node) (infoschema.InfoSchema, error) {
 				{
 					Name:   model.NewCIStr("c"),
 					Length: types.UnspecifiedLength,
-					Offset: 1,
+					Offset: 2,
```
Comment (shenli) [inline plan/logical_plan_test.go]: Why the offset changed?

File: plan/logical_plans.go [hunk 3]
```diff
@@ -179,6 +179,9 @@ type Selection struct {
 	// split into a list of AND conditions.
 	Conditions []expression.Expression
 
+	pushDownConditions []expression.Expression
```
Comment (shenli) [inline plan/logical_plans.go]: Add comments for the two members.

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -48,7 +55,7 @@ func (p *Projection) getPushedProp(prop *requiredProp) (*requiredProp, bool) {
 		case *expression.Column:
 			newCols = append(newCols, expr)
 		case *expression.ScalarFunction:
-			return nil, false
+			return newProp, false
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Is this a bug fix?

File: plan/new_physical_plan_builder.go [hunk 5]
```diff
@@ -145,30 +152,245 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+			orderedTask = cop.finishTask(p.basePlan.ctx, p.basePlan.allocator)
+		}
 		if orderedTask.cost() < task.cost() {
 			task = orderedTask
 		}
 	}
 	return task, p.storeTaskProfile(prop, task)
 }
+
+func (p *DataSource) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error) {
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Add a few comments in the following three functions.

File: plan/new_physical_plan_builder.go (hunk 4 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: no

File: plan/logical_plan_test.go (hunk 2 above)
Comment (hanfei1991) [inline plan/logical_plan_test.go]: the old offset is wrong

File: plan/task_profile.go [hunk 6]
```diff
@@ -241,32 +263,55 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
+			if is, ok := t.indexPlan.(*PhysicalIndexScan); ok {
+				conds = is.indexFilterConditions
 			}
-			if tableSel != nil {
-				t.finishIndexPlan()
-				tableSel.SetChildren(t.tablePlan)
-				t.tablePlan = tableSel
-				t.cst += float64(t.cnt) * cpuFactor
+		}
+		if t.tablePlan != nil && !t.addPlan2Index {
```
Comment (winoros) [inline plan/task_profile.go]: this branch and the above branch could connect with else?

File: plan/task_profile.go [hunk 7]
```diff
@@ -275,25 +320,24 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
 }
 
+func (sel *Selection) splitPushDownConditions() {
+	if len(sel.pushDownConditions)+len(sel.residualConditions) != 0 {
+		return
+	}
+	_, sel.pushDownConditions, sel.residualConditions = ExpressionsToPB(sel.ctx.GetSessionVars().StmtCtx, sel.Conditions, sel.ctx.GetClient())
+}
+
 // splitSelectionByIndexColumns splits the selection conditions by index schema. If some condition only contain the index
```
Comment (winoros) [inline plan/task_profile.go]: update comment

File: plan/task_profile.go [hunk 8]
```diff
@@ -275,25 +320,24 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
 // columns, it will be pushed to index plan.
-func (sel *Selection) splitSelectionByIndexColumns(schema *expression.Schema) (indexSel *Selection, tableSel *Selection) {
-	var tableConds []expression.Expression
-	var indexConds []expression.Expression
-	for _, cond := range sel.Conditions {
+func splitConditionsByIndexColumns(conditions []expression.Expression, schema *expression.Schema) (tableConds []expression.Expression, indexConds []expression.Expression) {
```
Comment (winoros) [inline plan/task_profile.go]: Seems it play the same role as `DetachIndexFilterCondition` in `refiner.go`, why that one considered pkCol and this not?

Comment (hanfei1991) [inline plan/task_profile.go]: I don't add pkCol to schema. It's my fault and I will add a TODO comment.

File: plan/task_profile.go (hunk 6 above)
Comment (hanfei1991) [inline plan/task_profile.go]: Well, it's clearer in this way.

**Comment by hanfei1991** [comment]
@coocood @shenli @winoros PTAL

File: plan/task_profile.go [hunk 9]
```diff
@@ -85,29 +85,39 @@ func attachPlan2TaskProfile(p PhysicalPlan, t taskProfile) taskProfile {
 // finishIndexPlan means we no longer add plan to index plan, and compute the network cost for it.
 func (t *copTaskProfile) finishIndexPlan() {
 	if t.addPlan2Index {
-		t.cst += float64(t.cnt) * netWorkFactor
+		t.indexPlan.SetSchema(expression.NewSchema()) // we only need the handle
```
Comment (coocood) [inline plan/task_profile.go]: If the index scan is single read or has selection pushed down, we should not set it to empty.

File: plan/task_profile.go [hunk 10]
```diff
@@ -149,20 +159,27 @@ func (p *Limit) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
+		}
 		cop = attachPlan2TaskProfile(pushedDownLimit, cop).(*copTaskProfile)
-		cop.setCount(pushedDownLimit.Count)
+		cop.setCount(float64(pushedDownLimit.Count))
 		profile = cop.finishTask(p.ctx, p.allocator)
 	}
-	profile = attachPlan2TaskProfile(p.Copy(), profile)
-	profile.setCount(p.Count)
+	np := p.Copy()
+	np.SetSchema(profile.(*rootTaskProfile).plan.Schema())
```
Comment (coocood) [inline plan/task_profile.go]: Can we make sure the `Limit` Plan has Schema set when it is created?

File: plan/task_profile.go [hunk 11]
```diff
@@ -34,16 +34,16 @@ type copTaskProfile struct {
 	indexPlan PhysicalPlan
 	tablePlan PhysicalPlan
 	cst       float64
-	cnt       uint64
+	cnt       float64
 	// addPlan2Index means we are processing index plan.
 	addPlan2Index bool
```
Comment (coocood) [inline plan/task_profile.go]: How about rename to `indexPlanFinished`?

File: plan/task_profile.go (hunk 10 above)
Comment (hanfei1991) [inline plan/task_profile.go]: We can make sure it.

File: plan/new_physical_plan_builder.go [hunk 12]
```diff
@@ -145,30 +154,296 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+	is := PhysicalIndexScan{
+		Table:       p.tableInfo,
+		TableAsName: p.TableAsName,
+		DBName:      p.DBName,
+		Columns:     p.Columns,
+		Index:       idx,
+	}.init(p.allocator, p.ctx)
+	statsTbl := p.statisticTable
+	rowCount := float64(statsTbl.Count)
+	sc := p.ctx.GetSessionVars().StmtCtx
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Need to set transaction’s readOnly value？

File: plan/new_physical_plan_builder.go [hunk 13]
```diff
@@ -145,30 +154,296 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+	if len(p.pushedDownConds) > 0 {
+		conds := make([]expression.Expression, 0, len(p.pushedDownConds))
+		for _, cond := range p.pushedDownConds {
+			conds = append(conds, cond.Clone())
+		}
+		is.AccessCondition, is.indexFilterConditions, is.accessEqualCount, is.accessInAndEqCount = DetachIndexScanConditions(conds, idx)
+		err = BuildIndexRange(sc, is)
+		if err != nil {
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Need to consider the truncated error?

File: plan/new_physical_plan_builder.go [hunk 14]
```diff
@@ -145,30 +154,296 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+			return nil, errors.Trace(err)
+		}
+		rowCount, err = statsTbl.GetRowCountByIndexRanges(sc, is.Index.ID, is.Ranges, is.accessInAndEqCount)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	} else {
+		rb := rangeBuilder{sc: p.ctx.GetSessionVars().StmtCtx}
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: s/p.ctx.GetSessionVars().StmtCtx/sc

File: plan/new_physical_plan_builder.go [hunk 15]
```diff
@@ -145,30 +154,296 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+	}
+	if !isCoveringIndex(is.Columns, is.Index.Columns, is.Table.PKIsHandle) {
+		// On this way, it's double read case.
+		copTask.tablePlan = PhysicalTableScan{Columns: p.Columns, Table: is.Table}.init(p.allocator, p.ctx)
+		copTask.tablePlan.SetSchema(p.schema)
+		var indexCols []*expression.Column
+		for _, col := range idx.Columns {
+			indexCols = append(indexCols, &expression.Column{FromID: p.id, Position: col.Offset})
+		}
+		// TODO: We should add pk column to index schema.
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: s/pk/PK

File: plan/new_physical_plan_builder.go [hunk 16]
```diff
@@ -145,30 +154,296 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
...
+			} else if i >= is.accessEqualCount {
+				matchProperty = false
+				break
+			}
+		}
+	}
+	if matchProperty && !prop.isEmpty() {
+		if prop.desc {
+			is.Desc = true
+			copTask.cst += 4 * copTask.cst
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Why the value is 4？ Please add a constant as this value. It uses two places.

File: plan/task_profile.go [hunk 17]
```diff
@@ -241,59 +270,82 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
-				tableSel.SetChildren(t.tablePlan)
-				t.tablePlan = tableSel
-				t.cst += float64(t.cnt) * cpuFactor
+		var indexConds, tableConds []expression.Expression
+		conds := sel.pushDownConditions
+		if t.indexPlanFinished {
+			if ts, ok := t.tablePlan.(*PhysicalTableScan); ok {
+				conds = ts.tableFilterConditions
 			}
+		} else if is, ok := t.indexPlan.(*PhysicalIndexScan); ok {
```
Comment (coocood) [inline plan/task_profile.go]: Those bunch of code is too complex.
Can we just check `t.indexPlanFinished` only once?

File: plan/new_physical_plan_builder.go (hunk 16 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Desc Scan Is Five times lower than Scan.

File: plan/new_physical_plan_builder.go (hunk 13 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: I think we needn't, can you give some explames that we can ignore truncate err?

File: plan/task_profile.go (hunk 17 above)
Comment (hanfei1991) [inline plan/task_profile.go]: I didn't find a good way.

File: plan/new_physical_plan_builder.go (hunk 12 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Why do we need read only value ?

Comment (zimulala) [inline plan/new_physical_plan_builder.go]: The old code has it. But I find we don't use it.
Ignore this comment.

**Comment by hanfei1991** [comment]
@coocood @zimulala PTAL

File: plan/new_physical_plan_builder.go (hunk 13 above)
Comment (coocood) [inline plan/new_physical_plan_builder.go]: Truncate error is handled by `sc`.

File: plan/task_profile.go [hunk 18]
```diff
@@ -241,59 +270,82 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
+		var indexConds, tableConds []expression.Expression
+		conds := sel.pushDownConditions
+		if t.indexPlanFinished {
+			if ts, ok := t.tablePlan.(*PhysicalTableScan); ok {
+				conds = ts.filterCondition
 			}
+		} else if is, ok := t.indexPlan.(*PhysicalIndexScan); ok {
+			conds = is.filterCondition
+		}
+		if t.tablePlan != nil && !t.indexPlanFinished {
```
Comment (coocood) [inline plan/task_profile.go]: Think the first `if` use single condition is better, like
```
if t.indexPlanfinished {
...
} else if t.tablePlan != nil {
...
} else {
...
}
```
This way, every condition is checked only once.

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

File: plan/task_profile.go [hunk 19]
```diff
@@ -56,82 +57,94 @@ func (t *copTaskProfile) cost() float64 {
...
 }
 
 // finishIndexPlan means we no longer add plan to index plan, and compute the network cost for it.
 func (t *copTaskProfile) finishIndexPlan() {
-	if t.addPlan2Index {
-		t.cst += float64(t.cnt) * netWorkFactor
-		t.addPlan2Index = false
+	if !t.indexPlanFinished {
+		if t.tablePlan != nil {
+			t.indexPlan.SetSchema(expression.NewSchema()) // we only need the handle
```
Comment (hanfei1991) [inline plan/task_profile.go]: In case of double reading, we only need to read hanlde from index.

File: plan/task_profile.go [hunk 20]
```diff
@@ -241,59 +270,82 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
-			}
-			if tableSel != nil {
-				t.finishIndexPlan()
-				tableSel.SetChildren(t.tablePlan)
-				t.tablePlan = tableSel
-				t.cst += float64(t.cnt) * cpuFactor
+		var indexConds, tableConds []expression.Expression
+		conds := sel.pushDownConditions
+		if t.indexPlanFinished {
+			if ts, ok := t.tablePlan.(*PhysicalTableScan); ok {
```
Comment (zimulala) [inline plan/task_profile.go]: Do we need to consider the value of `t.tablePlan` is nil?

Comment (hanfei1991) [inline plan/task_profile.go]: No, we needn't. If the table plan is nil, the task has been finished and it can never be a copTask.

Comment (zimulala) [inline plan/task_profile.go]: Got it.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**